examples/petsc/setup.h

```diff
@@ -0,0 +1,881 @@
+
+  ierr = DMPlexGetHeightStratum(dm, 0, &cStart, &cEnd); CHKERRQ(ierr);
+  nelem = cEnd - cStart;
+
+  CeedElemRestrictionCreateIdentity(ceed, nelem, Q*Q*Q, nelem*Q*Q*Q, ncompu,
+                                    &Erestrictui); CHKERRQ(ierr);
+  CeedElemRestrictionCreateIdentity(ceed, nelem, Q*Q*Q, nelem*Q*Q*Q,
+                                    qdatasize, &Erestrictqdi); CHKERRQ(ierr);
+  CeedElemRestrictionCreateIdentity(ceed, nelem, Q*Q*Q, nelem*Q*Q*Q, dim,
+                                    &Erestrictxi); CHKERRQ(ierr);
+
+  // Element coordinates
+  ierr = DMGetCoordinatesLocal(dm, &coords); CHKERRQ(ierr);
+  ierr = VecGetArrayRead(coords, &coordArray); CHKERRQ(ierr);
+  ierr = DMGetSection(dmcoord, &section); CHKERRQ(ierr);
+
+  CeedElemRestrictionCreateVector(Erestrictx, &xcoord, NULL);
+  CeedVectorSetArray(xcoord, CEED_MEM_HOST, CEED_COPY_VALUES,
+                     (PetscScalar *)coordArray);
+  ierr = VecRestoreArrayRead(coords, &coordArray); CHKERRQ(ierr);
+
+  // Create the persistent vectors that will be needed in setup and apply
+  CeedInt nqpts;
+  CeedBasisGetNumQuadraturePoints(basisu, &nqpts);
+  CeedVectorCreate(ceed, qdatasize*nelem*nqpts, &qdata);
+  CeedVectorCreate(ceed, xlsize, &xceed);
+  CeedVectorCreate(ceed, xlsize, &yceed);
+
+  // Create the Q-function that builds the operator (i.e. computes its
+  // quadrature data) and set its context data
+  CeedQFunctionCreateInterior(ceed, 1, bpOptions[bpChoice].setupgeo,
+                              bpOptions[bpChoice].setupgeofname, &qf_setupgeo);
+  CeedQFunctionAddInput(qf_setupgeo, "dx", dim*dim, CEED_EVAL_GRAD);
+  CeedQFunctionAddInput(qf_setupgeo, "weight", 1, CEED_EVAL_WEIGHT);
+  CeedQFunctionAddOutput(qf_setupgeo, "qdata", qdatasize, CEED_EVAL_NONE);
+
+  // Set up PDE operator
+  CeedInt inscale = bpOptions[bpChoice].inmode==CEED_EVAL_GRAD ? dim : 1;
+  CeedInt outscale = bpOptions[bpChoice].outmode==CEED_EVAL_GRAD ? dim : 1;
+  CeedQFunctionCreateInterior(ceed, 1, bpOptions[bpChoice].apply,
+                              bpOptions[bpChoice].applyfname, &qf_apply);
+  CeedQFunctionAddInput(qf_apply, "u", ncompu*inscale,
+                        bpOptions[bpChoice].inmode);
+  CeedQFunctionAddInput(qf_apply, "qdata", qdatasize, CEED_EVAL_NONE);
+  CeedQFunctionAddOutput(qf_apply, "v", ncompu*outscale,
+                         bpOptions[bpChoice].outmode);
+
+  // Create the operator that builds the quadrature data for the operator
+  CeedOperatorCreate(ceed, qf_setupgeo, NULL, NULL, &op_setupgeo);
+  CeedOperatorSetField(op_setupgeo, "dx", Erestrictx, CEED_TRANSPOSE,
+                       basisx, CEED_VECTOR_ACTIVE);
+  CeedOperatorSetField(op_setupgeo, "weight", Erestrictxi, CEED_NOTRANSPOSE,
+                       basisx, CEED_VECTOR_NONE);
+  CeedOperatorSetField(op_setupgeo, "qdata", Erestrictqdi, CEED_NOTRANSPOSE,
+                       CEED_BASIS_COLLOCATED, CEED_VECTOR_ACTIVE);
+
+  // Create the diff operator
```

Comment:
@jeremylt why `diff` here? Couldn't this represent any operator according to the BP selected?

Reply:
Typo